### ISSUE TITLE
RandomSampler iter methods cause failure with huge dataset size

### ISSUE THREAD

**Issue by tobyclh** [pytorch/pytorch#11691]
## Issue description
I created a video dataset where which each data includes only a few frames, 
including all the permutation, this makes the total size of the dataset stupidly huge (higher than 1e9).
Whenever I try to set shuffle=True in dataloader, the system kills the process for taking too much resource.
After some investigation, it is 
`iter(torch.randperm(len(self.data_source)).tolist())`
in the iter method of random sampler causing this failure.
That leaves me with 2 solutions.
1. Use shuffle=False (not ideal)
2. Create a custom sampler with numpy (which in my testing, is very slow but gets the job done)

## Code example
```
dataset = SpeechAnimationDataset(opt)

loader = DataLoader(dataset, batch_size=5, num_workers=2) #works
loader = DataLoader(dataset, batch_size=5, num_workers=2, shuffle=True) #fails after long process
data_iter = iter(loader)
for i in tqdm(range(500)):
    data = next(data_iter)
```

Working numpy replacement 
```
class RandomSampler(Sampler):
    r"""Samples elements randomly, without replacement.

    Arguments:
        data_source (Dataset): dataset to sample from
    """

    def __init__(self, data_source):
        self.data_source = data_source

    def __iter__(self):
        arr = np.arange(len(self.data_source))
        np.random.shuffle(arr)
        return iter(arr)

    def __len__(self):
        return len(self.data_source)
```
I also tried to remove the .tolist() that didn't seem necessary, but that didn't fix anything. 
I don't see pytorch have much numpy code in the repo, so I guess the method I propose is not ideal, but I think this could be useful to some who have the same problem.
Thanks!

**Comment by SsnL** [comment]
If you are randomly permuting `1e9` numbers, you are going to be slow.

**Comment by tobyclh** [comment]
Thanks for the response.
But I didn't mean slow, but actual crashing.
The same behavior doesn't occur when I use numpy.
Is there something we can do to enhance this?

**Comment by tobyclh** [comment]
@fmassa 
Thanks again for the kind and helpful reply.
I agree with you, and that's the way I went down.

For future reference who stumble on this,
instead of referencing the normal random sampler, 
try to have a look at how BatchSampler handles large list with yield statement to generate random numbers on the fly.

**Comment by codyohl** [comment]
@tobyclh you might also want to look at this to solve this type of problem without having duplicate samples: https://fossies.org/linux/tensorflow/tensorflow/core/kernels/random_index_shuffle.h
